Libraries/LibWeb/CSS/StyleComputer.cpp

```diff
@@ -1042,23 +1058,60 @@ void StyleComputer::cascade_declarations(
+            auto property_value = property.value;
+
+            if (property_value->is_unresolved())
+                property_value = Parser::Parser::resolve_unresolved_style_value(Parser::ParsingParams { element.document() }, element, pseudo_element, property.property_id, property_value->as_unresolved());
+
+            if (property_value->is_guaranteed_invalid()) {
+                // https://drafts.csswg.org/css-values-5/#invalid-at-computed-value-time
+                // When substitution results in a property’s value containing the guaranteed-invalid value, this makes the
+                // declaration invalid at computed-value time. When this happens, the computed value is one of the
+                // following depending on the property’s type:
+
+                // -> The property is a non-registered custom property
+                // -> The property is a registered custom property with universal syntax
+                // FIXME: Process custom properties here?
+                if (false) {
+                    // The computed value is the guaranteed-invalid value.
+                }
+                // -> Otherwise
+                else {
+                    // Either the property’s inherited value or its initial value depending on whether the property is
+                    // inherited or not, respectively, as if the property’s value had been specified as the unset keyword.
+                    property_value = CSSKeywordValue::create(Keyword::Unset);
+                }
+            }
+
+            if (property.property_id == PropertyID::All) {
+                set_all_properties(cascaded_properties, element, pseudo_element, property_value, m_document, &declaration, cascade_origin, important, layer_name);
                 continue;
             }
 
-            auto property_value = property.value;
-            if (property.value->is_unresolved())
-                property_value = Parser::Parser::resolve_unresolved_style_value(Parser::ParsingParams { document() }, element, pseudo_element, property.property_id, property.value->as_unresolved());
-            if (!property_value->is_unresolved())
-                set_property_expanding_shorthands(cascaded_properties, property.property_id, property_value, &declaration, cascade_origin, important, layer_name);
+            // NOTE: This is a duplicate of set_property_expanding_shorthands() with some extra checks.
+            for_each_property_expanding_shorthands(property.property_id, property_value, AllowUnresolved::No, [&](PropertyID longhand_id, CSSStyleValue const& longhand_value) {
+                // If we're a PSV that's already been seen, that should mean that our shorthand already got
+                // resolved and gave us a value, so we don't want to overwrite it with a PSV.
+                if (seen_properties.contains(longhand_id) && property_value->is_pending_substitution())
+                    return;
+                seen_properties.set(longhand_id);
```

Comment:
This feels a bit confusing. Can we add some tests to verify that e.g `margin` and `margin-top` with PSV in mixed order actually resolves correctly?